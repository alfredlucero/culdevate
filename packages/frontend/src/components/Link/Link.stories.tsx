import React from "react";
import { BrowserRouter } from "react-router-dom";
import { storiesOf } from "@storybook/react";
import Link from "./index";

storiesOf("Common/Link", module)
  .add("Internal", () => (
    <BrowserRouter>
      <Link href="#" type="internal">
        Internal Link
      </Link>
    </BrowserRouter>
  ))
  .add("External", () => (
    <Link href="#" type="external">
      External Link
    </Link>
  ));
