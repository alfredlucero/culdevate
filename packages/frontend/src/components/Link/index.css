.link {
  @apply font-bold;
  @apply text-blue-500;
}

.link:hover {
  @apply text-blue-800;
}
