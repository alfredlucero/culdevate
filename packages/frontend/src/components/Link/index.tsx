import React from "react";
import { Link as RouterLink } from "react-router-dom";
import cn from "classnames";
import { CommonProps } from "../commonProps";

interface LinkProps extends CommonProps {
  href: string;
  type: "internal" | "external";
  children: React.ReactNode;
}

const linkBaseClasses = ["font-bold", "text-teal-500", "hover:text-teal-800"];

const Link: React.FC<LinkProps> = ({ href, type, children, testId = "", className = "", ...passThroughProps }) => {
  // Internal links utilize the React Router links to navigate
  // to our SPA routes
  // These should be relative paths i.e. /login
  if (type === "internal") {
    return (
      <RouterLink
        to={href}
        className={cn(linkBaseClasses, className)}
        {...(testId !== "" ? { "data-testid": testId } : {})}
        {...passThroughProps}
      >
        {children}
      </RouterLink>
    );
  }

  // External links are normal <a> tags that are meant to go outside of our app domain
  // and open in another tab
  // These should be full external links i.e. https://domain.com
  if (type === "external") {
    return (
      <a
        href={href}
        className={cn(linkBaseClasses, className)}
        target="_blank"
        rel="noopener noreferrer"
        {...(testId !== "" ? { "data-testid": testId } : {})}
        {...passThroughProps}
      >
        {children}
      </a>
    );
  }

  return null;
};

export default Link;
