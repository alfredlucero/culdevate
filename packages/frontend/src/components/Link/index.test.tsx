import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Link from "./index";

describe("<Link />", () => {
  test("should render internal link without error", () => {
    const { container } = render(
      <BrowserRouter>
        <Link href="/login" type="internal" className="extra-link-class" testId="internalLinkId">
          Internal Link
        </Link>
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render external link without error", () => {
    const { container } = render(
      <BrowserRouter>
        <Link href="https://external.com" type="external" className="extra-link-class" testId="externalLinkId">
          External Link
        </Link>
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });
});
