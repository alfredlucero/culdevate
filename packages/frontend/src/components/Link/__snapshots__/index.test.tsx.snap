// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should render external link without error 1`] = `
<div>
  <a
    class="font-bold text-teal-500 hover:text-teal-800 extra-link-class"
    data-testid="externalLinkId"
    href="https://external.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    External Link
  </a>
</div>
`;

exports[`<Link /> should render internal link without error 1`] = `
<div>
  <a
    class="font-bold text-teal-500 hover:text-teal-800 extra-link-class"
    data-testid="internalLinkId"
    href="/login"
  >
    Internal Link
  </a>
</div>
`;
